### ISSUE TITLE
Pattern guards

### ISSUE THREAD

**Issue by krame505** [melt-umn/silver#373]
An annoying issue that occasionally shows up in match rules is when a guard expression depends on some value that is computed from pattern variables, while the same computation is required on the right side of the rule, requiring the computation to be repeated for purely syntactic reasons.  This most often arises in the context of term rewriting.  

Haskell allows for "pattern guards", i.e. guards that match an arbitrary expression, potentially involving variables in the main pattern, against an additional pattern: https://wiki.haskell.org/Pattern_guard

This would allow us to write match rules of the form (using syntax proposed by Eric)
```haskell
| var(n) when lookup(n, top.env) matches just(e) -> e
```

Note that exactly copying Haskell's syntax here would yield
```haskell
| var(n) when just(e) <- lookup(n, top.env) -> e
```
which is hard to parse and (in my opinion) hard to read as well.

Haskell allows for an arbitrary list of guards to be chained together, separated by commas.  This means that someone could also write
```haskell
| var(n) when lookup(n, top.env) matches just(e), e.somePredicate -> e
```

Haskell also allows `let`-bindings as guards in the case that a value computed from a pattern variable is needed more than once, for example the original match rule from above could be written without using `... matches ...` as:
```haskell
| var(n) when let res = lookup(n, top.env), res.isJust -> res.fromJust
```

What combination of these features would be useful/sufficient?  I believe only allowing for a single pattern guard (as an alternative to a Boolean guard) would cover the majority of cases, but would we want to potential still support arbitrary chaining of guards?

**Comment by ericvanwyk** [comment]
Any documentation written for this nice feature?

**Comment by ericvanwyk** [comment]
yes, reopening this is a bit of a passive-aggressive move... but we should write something about this in the docs.
